main_test.go

```diff
@@ -1,62 +1,198 @@
 package main
 
 import (
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
 	"testing"
 
-	. "github.com/onsi/ginkgo"
-	. "github.com/onsi/gomega"
+	"github.com/stretchr/testify/assert"
 )
 
-func TestMain(t *testing.T) {
-	RegisterFailHandler(Fail)
-	RunSpecs(t, "main")
+func GetTestConfig() Config {
+
+	originUrl, _ := url.Parse("http://origin.url")
+	proxyUrl, _ := url.Parse("http://proxy.url")
+	ssoUrl, _ := url.Parse("http://sso.url")
+
+	config := Config{
+		OriginURL:       originUrl,
+		ProxyURL:        proxyUrl,
+		SSOURL:          ssoUrl,
+		SSOSecret:       "secret",
+		AllowAll:        false,
+		AllowGroups:     "",
+		BasicAuth:       "",
+		Whitelist:       "",
+		UsernameHeader:  "username-header",
+		GroupsHeader:    "groups-header",
+		Timeout:         10,
+		SRVAbandonAfter: 600,
+		LogRequests:     false,
+	}
+
+	return config
 }
 
-var _ = Describe("parseCookie", func() {
-	var (
-		signed, secret string
+type SSOOptions struct {
+	URL    string
+	Secret string
+	Nonce  string
+	Groups string
+	Admin  bool
+}
 
-		parsedUsername string
-		parsedGroup    string
-		parseError     error
-	)
+func GetSSOOptions(url string, secret string) SSOOptions {
+	options := SSOOptions{
+		URL:    url,
+		Secret: secret,
+		Nonce:  "",
+		Groups: "",
+		Admin:  false,
+	}
+	return options
+}
 
-	JustBeforeEach(func() {
-		parsedUsername, parsedGroup, parseError = parseCookie(signed, secret)
-	})
+func GetTestSSOURL(options SSOOptions) string {
 
-	Context("when verifying with an invalid secret", func() {
-		BeforeEach(func() {
-			secret = "secretbar"
-			signed = signCookie("user,group", "secretfoo")
-		})
+	u, _ := url.Parse(options.URL)
 
-		It("fails", func() {
-			Expect(parseError).To(HaveOccurred())
-		})
-	})
+	sso := fmt.Sprintf("admin=%v&username=sam&groups=%v&nonce=", options.Admin, options.Groups)
 
-	Context("when verifying with an invalid payload", func() {
-		BeforeEach(func() {
-			secret = "mysecret"
-			signed = signCookie("user,group", secret) + "garbage"
-		})
+	var nonce = ""
+	if options.Nonce != "" {
+		nonce = options.Nonce
+	} else {
+		nonce = addNonce("http://some.url/")
```

Comment:
looks like there is a Cleanup now, added.

Reply:
I still think readability would be improved with split functions.  Even in Go, we try to avoid mixing unrelated concerns within a single function.  (You could hook the `Cleanup` from the suggested `RegisterNonce` function; accepting a `*testing.T` as a parameter in that small function will act as an even clearer indicator that the function imparts unusual side effects, with the URL builder remaining pure.)

The way that `sso` string was partially constructed was also a little unusual.  (As it first exists on line 65, it is incomplete.  I finished reading that line and thought, _Isn't something missing here?_)  The suggested solution does not use partially constructed values.